Fixed #31034 -- Added a navigation sidebar to the admin

knyghty opened the pull request:
https://code.djangoproject.com/ticket/31034

I'm not altogether sure about my approach here. This seems to be the way that is the least disruptive to me, (almost) all the CSS is in a new file so it shouldn't affect users that have it turned off.

Really I think something like a drawer that slides from the side and covers some of the page would be better, as even with it only enabled on the largest breakpoint it can be a little crowded at the minimum size (1024px). Open to suggestions here. Also couldn't find a combination of `height`s and `min-height`s that would get the sidebar to cover the whole height of the screen without breaking something.

I decided to default to on since it seems like the majority of users would want that, and it's easy enough to turn off if upgrading and not wanting it.

Tested with latest Chrome, Firefox, Safari. Could use some eyes on other browsers.

Will add tests/docs after getting some feedback.

Screens:

15" Macbook:
<img width="1440" alt="Screenshot 2019-12-05 at 14 22 05" src="https://user-images.githubusercontent.com/3871354/70239136-06213300-176b-11ea-81f5-cb34656f752a.png">

At 1025px, the smallest size the sidebar will show up (a little crowded, but didn't want to add another breakpoint):
<img width="1029" alt="Screenshot 2019-12-05 at 14 24 00" src="https://user-images.githubusercontent.com/3871354/70239203-2b15a600-176b-11ea-981e-f292dbb06f70.png">

olivierdalang commented:
Hi, thank you very much for the work on this, it will be very much appreciated here. To get more comments, maybe you could add screenshots (both here and on the mailing list) ? 

I had a quick look on desktop :

**breadcrumbs bar** : currently it conflicts both visually and functionnaly.
I'd suggest putting the bar to the right of the menu instead of the top. I'd also suggest to remove the bar altogether for the basic admin views, as thanks to the menu, most breadcrumbs would be useless. The only point where the breadcrumbs bar makes sense is to get back to the main instance from some custom instance views (e.g. admin history).

**add buttons** : is it possible to have the "add" button in the menu as in the admin_site view ? the "change" button is useless though, as it's the same than the link

**folding** : for projects with many apps and apps with many models, the menu becomes really long. it would be great to be able to fold apps.

**ordering/filtering** : do you plan to add some ability to order/filter items in the menu ? It's also something that's very much missing to have good navigation in the admin (even on the admin_index view)

knyghty commented:
Thanks for the feedback, @olivierdalang 

Yep, agreed, I will add some screenshots here soon.

For your suggestions:

**breadcrumbs bar**: yeah it does clash visually somewhat. I'll take another look at that. I think you're right that it's redundant for a lot of pages, but I also worry that it will feel weird to only have breadcrumbs on a few specific pages. I feel like I want to leave it there, I think that having more options is better than fewer, and also that it reduces the chance of bugs if I mess with less code.

**add buttons**: good idea. While you can get to the add page easily from the list, I think it's probably okay to have an `[add]` link - I'll see how it looks anyway.

**folding**: I think this is a great idea, but a little impractical to implement right now. It'd be a great improvement, but for now it's a little hard to figure out how it would work. The admin just renders templates, so we can't easily keep the state of what's folded/not between reloads, at least without adding a fair amount of code (local storage, perhaps?). Another option is just to have everything folded except what is "active" but this seems a little useless to me. I think for now if you have so many apps/models it would make the sidebar too long, perhaps a better solution (if possible) is to make more use of groups and give people access only to the apps/models they need.

**ordering/filtering**: definitely out of scope for what I want to do with this PR, but might make a nice improvement after/if this is merged.

knyghty commented:
Added tests. It could probably use more, but need some direction on what's useful here.

MarkusH commented:
retest this please

MarkusH commented:
test this please

felixxm commented:
buildbot, test on selenium.

felixxm commented:
@knyghty Please take a look at failing test and change the `ADMIN_ENABLE_NAV_SIDEBAR` setting to an admin site attribute.

knyghty commented:
@felixxm Thanks. I've moved this to be an `AdminSite` attribute. I'm having some issues with the test, I can't figure out what the issue is, other than something is apparently causing a number to be bumped by one. It seems unrelated to my code, but I don't know why it occurs here. I asked on `#django-dev` without any luck, so I'd appreciate any help debugging it.

knyghty commented:
@felixxm I've fixed the test by just changing the value in it.

I took a look at the history of that test and it looks like the number has changed before and the test updated, so I just changed it here as well so it'll pass. I tried messing around with some inlines / formets in the admin and they seem to work okay for me. Hopefully that's ok.

olivierdalang commented:
> breadcrumbs bar: (...) I also worry that it will feel weird to only have breadcrumbs on a few specific pages.

That's already the case : breadcrumbs don't appear on the admin index. It would be so much better without it :-)

sarendsen commented:
Great job. But 100% height on the sidebar would really finish it.
Here's my stab at it..

nav_sidebar.css

```css
#container {
    min-height: 100vh;
}

.login #container {
    min-height: auto;
}

#content {
    float: left;
    width: calc(100% - 280px);
    padding-left: 240px;
}

.dashboard #content {
    width: calc(100% - 580px);
}

.nav-sidebar {
    background: #f8f8f8;
    height: 100vh;
    font-size: 13px;
    position: fixed;
    width: 200px;
}

.nav-app {
    margin: 20px 0;
}

.nav-app:first-child {
    margin-top: 0;
}

.nav-app-name {
    background: #79aec8;
    display: block;
    font-size: 12px;
    letter-spacing: 0.5px;
    margin-bottom: 5px;
    padding: 8px;
    text-transform: uppercase;
}

.nav-app-name:link, .nav-app-name:visited {
    color: #fff;
}

.nav-app-name:hover {
    text-decoration: underline;
}

.nav-model {
    font-weight: bold;
    padding: 5px 10px;
}

@media (max-width: 1024px) {
    .nav-sidebar {
        display: none;
    }
}
```

ps: When testing this PR I noticed the following:

- width of the content container did not adjusted to smaller viewports (max 1024px) on several views like the listing. (all browsers).
- And login form fields where not 100% width on Safari.

sarendsen commented:
Another suggestion, line up the sidebar padding with the header/ breadcrumb. (might be better to move the breadcrumb like discussed above)

<img width="663" alt="Screenshot 2020-03-08 at 14 06 08" src="https://user-images.githubusercontent.com/168401/76163395-0227f700-6146-11ea-88cb-ad03b9721304.png">

knyghty commented:
@olivierdalang I've tried to add buttons for add but without success, the space limitation makes it quite hard to make something non-ugly. If you have a suggestion it'd be good to hear it, but I don't think it's too important.

knyghty commented:
@sarendsen wow, thanks for that. I had tried to get the full-height working initially but my CSS-foo was not so good. This is working well. I've also fixed the login form issue (which I'd fixed before but maybe I broke it again somehow).

The problem I see with changing the alignment is that because of the space restrictions stemming from fitting the sidebar into the original design, I don't want to indent the navbar content all the way to match the heading indentation. I also don't want to move the heading indentation back as I feel like I'd be messing with the original design intent. Personally I'm okay with how it is, but I'm open to tuning it somewhat.

> width of the content container did not adjusted to smaller viewports (max 1024px) on several views like the listing. (all browsers).

I'm not completely sure what you mean by this, but the smallest breakpoint is 1024px and I didn't want to add any new breakpoints just for this feature, it wouldn't feel very consistent to me.

olivierdalang commented:
> I've tried to add buttons for add but without success, the space limitation makes it quite hard to make something non-ugly. If you have a suggestion it'd be good to hear it, but I don't think it's too important.
@knyghty No good suggestion, but agree it's not too important.

I'm not exactly sure what will happen to the breadcrumbs bar though ? @sarendsen's screens shows it once with the breadcrumbs bar, and once without it.

If not changed, I'd like to insist once more one that :-) Having duplicated links really defeats the improvement in navigation clarity IMO. Also the sidebar takes up more screen space, so at least we should reclaim what we can !

Thanks for the great work 👍

knyghty commented:
@olivierdalang I haven't forgotten about the breadcrumbs issue, would just like some feedback on that from the core team, or someone with more UX knowledge at least.

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2837,6 +2837,13 @@ Templates can override or extend base admin templates as described in
+
+    .. versionadded:: 3.1
+
+    A boolean value that determines whether to show the navigation sidebar.
```
`...the navigation sidebar on larger screens.` I think.

carltongibson commented on docs/releases/3.1.txt:
```diff
@@ -39,6 +39,9 @@ Minor features
 
 * Filters in the right sidebar of the admin changelist view now contains a link
   to clear all filters.
+* The admin now has a sidebar for easier navigation. It is enabled by default
```
Again here, I think I'd add a `on larger screens`. (Or similar.)

carltongibson commented:
Hi @knyghty 

Found another issue: 

![Screenshot 2020-04-15 at 10 28 56](https://user-images.githubusercontent.com/64686/79315968-fdecb780-7f03-11ea-9ba7-d7d75cbbf8bf.png)

Messages are overlaid. 

(Admin == lots of variations to check.)

goetzb commented:
Thanks for posting the link on the mailing list. I like the idea of a sidebar a lot. It would probably be good if we could highlight the current page/model in the sidebar in some way, like by adding an `active` class with different colours or something similar, if possible, to make a bit clearer in the sidebar itself where the user currently is.

In my view, especially without breadcrumbs navigation, the different capitalisation of the model name in the sidebar ("Model") and in the heading ("Select model to change") in the main content section could be a bit confusing too.

carltongibson commented:
@goetzb Are you able to help with the design/implementation there?

jaylynnconley12 commented:
ummmm

goetzb commented:
> @goetzb Are you able to help with the design/implementation there?

@carltongibson I could perhaps have a closer look at the current code, yes. But to provide more detailed suggestions, I will first need set up a new environment with a development version of Django (I believe that's all documented somewhere, so I just need to read up on it and do it). Not sure I'll get around doing that today, but should manage to do it tomorrow.

goetzb commented:
I've now made a few changes to the code from @knyghty in https://github.com/goetzb/django/commit/b54ac9f71803fcd505a4901739cbb1f0ad8664cf

* Changed the sidebar menu markup to nested lists (seems best practice and also good for accessibility)
* Added active classes to CSS, improved accessibility with `aria-current="page"`
* Added visual highlighting for active app and model sidebar items
* Added a small border above the sidebar if the breadcrumbs navigation is present before the `.nav-sidebar` in the HTML
* Added hidden "Main Menu" header for accessibility (forgot to change the text to "Sidebar Menu")
* Fixed the message position when the `.nav-sidebar` immediately before `ul.messagelist` in HTML (highlighted by @carltongibson - I've made that change in `base.css`, but it should probably also go into `nav_sidebar.css`)

<img width="361" alt="Screenshot 2020-04-16 12 25 56" src="https://user-images.githubusercontent.com/1097522/79450873-7f644880-7fdd-11ea-8b93-6b160a589214.png">

Looking at this screenshot, I find the positioning of the text in the header and the sidebar not ideal - maybe we should now standardise the `padding-left` for the header, the breadcrumbs and the sidebar?

olivierdalang commented:
What about moving the breadcrumbs next to the page title like this :

![image](https://user-images.githubusercontent.com/1894106/79452273-2c43c300-7fe8-11ea-93dd-c33c69b88d95.png)

That would alleviate this "duplicated navigation" effect, and would also provide more stable structure between pages with and without breadcrumbs (the homepage has no breadcrumbs).

goetzb commented:
@olivierdalang Good idea, yes. I just wanted to touch as little outside the sidebar as possible. Not sure to what extent people customise the breadcrumbs in a way that moving it around would mess up their admin. But if this is a change that's acceptable to make, then yes, this would make the navigation overall a bit more compact.

carltongibson commented:
Hmmm. I'm a bit worried we take the design and break it moving bits without an overall plan. (I shall try and workout who did the design when the admin was updated last...)

carltongibson commented:
Hi @elky — The modern flat theme for the admin is your creation I believe. Thanks for that work. I'm a big fan. 

We're looking to add a sidebar here. You'll see the discussion. Could I ask you to review/comment if you have the capacity?

goetzb commented:
Thanks @carltongibson for your feedback, I agree that it's a good idea to check with @elky.

I've just made some tweaks in my branch on https://github.com/goetzb/django/commit/e170669976a0e323bb550a6701303fc59bb6d838

* Keeps active highlight for subpages (edit/create/delete pages of a model)
* Adds the left border highlight on the app navigation item for model pages as visual feedback that the active model is within that app

On other issue I've noticed is that the sidebar is fixed, but the header is not. So the header and breadcrumbs scroll out of the screen, leaving a white gap, while the sidebar remains visible.

<img width="301" alt="Screenshot 2020-04-16 16 16 18" src="https://user-images.githubusercontent.com/1097522/79474424-45f00500-7ffe-11ea-8578-e057b7399dc5.png">

I think having the header and the sidebar and breadcrumbs all `position: fixed;` would make sense, but is also potentially another breaking change.

A less intrusive option would be to position a box on the left in the background with the same width as the sidebar, so at least the area on the left is for example all grey (or blue) and not white and grey. Also, I'm currently trying to enable scrolling in the sidebar, which turns out trickier than it seems at first.

elky commented:
Hi everyone,

Thanks to @carltongibson for invitation. I like the idea of sidebar -- this is what I personally would like to use.

In terms of UX I would make this sidebar as a classic toggle sliding panel.

Advantages:
- We can use it globally (for all admin pages except dashboard and login page obviously)
- It saves a lot of space. I have some projects that have tons of columns in Change List view and pretty complicated inlines in Change Form view, so space there is really important
- It's available not only for desktops
- It's optional with toggler near to admin title

Disadvantages:
- One extra click to get sidebar content
- Few JS lines to make it work (+ "click outside" behaviour)

I also think that sidebar sections ("modules") should be the same style as on dashboard page, just to make it consistent.

Here's my quick demo I created in Chrome inspector 

<img width="1440" alt="Screenshot 2020-04-16 at 18 41 22" src="https://user-images.githubusercontent.com/209663/79478502-2bc12180-8014-11ea-9b63-bba2ad5d9104.png">

knyghty commented:
@elky this is exactly the sort of iterative improvement I hoped could be done after we merge this, but if we can do it now, all the better 👍 

I think this combined with @goetzb's active class would be good additions.

Is it possible / reasonable / good to remember the state of the sidebar between page loads? So we can keep the sidebar open if we're navigating a lot. This might not be too reasonable if it's covering a lot of content, though. I wonder if instead of having it slide over the content, if it could slide and push the content smaller?

elky commented:
> good to remember the state of the sidebar

It makes sense, yeah. We can use cookies or localStorage for this

> slide and push the content

I like it. We don't need to show and manage backdrop in this case. And most part of shifted content is still available. 

Possible disadvantages:
- It may break third-party apps which wrapped `#content` container
- We have to be sure there is nothing outside `#content` div

knyghty commented:
For the disadvantages, I think this is okay because the sidebar can be opted out of. Local storage sounds good to me since we're supporting only modern browsers now, and it's not a critical part of the feature if it doesn't work.

Another disadvantage is that you may crush the content too much on mobile, perhaps on very small screens it should cover, though I'm not sure how pracitcal it is.

It is a little beyond my ability, at least to do it quickly and comfortably, though. @elky, do you have any time to branch off this PR so I can merge that into here - I'm happy to tweak from that point and incorporate the active class changes.

goetzb commented:
Thank you for the input @elky - like @knyghty, I also think having an option to display the sidebar "permanently" would be a good idea.

I've now adjusted the design of my sidebar suggestions in https://github.com/goetzb/django/commit/5fd84876979b872729cf993a3b50a8a0bd5dfaf1 and have taken in your recommendations to align it with the dashboard design. But I haven't touched the header or added a menu icon.

knyghty commented:
@goetzb Thanks for your work on this 👍 

I've pulled your changes into this PR. Somehow the messages were still broken for me, I'm not sure why I had a different experience. I also found the fixed navbar didn't quite work right for me, so I've changed this to use `position: sticky`. This lets the sidebar stick to the top of the screen without using `fixed` and a background element (see screen). It's not supported on IE11 but it's not essential, the UX is still reasonable without it. This has also allowed me to clean up the code slightly. I also brought the messages CSS change into the sidebar CSS, as we don't need to send it to the user if the sidebar is disabled.

@carltongibson I've also addressed your docs concerns.

<img width="692" alt="Screenshot 2020-04-17 at 18 22 13" src="https://user-images.githubusercontent.com/3871354/79592262-0a2a6d80-80da-11ea-8927-070768805a18.png">
<img width="806" alt="Screenshot 2020-04-17 at 18 22 37" src="https://user-images.githubusercontent.com/3871354/79592280-0f87b800-80da-11ea-9183-9d931d64b528.png">

knyghty commented:
I noticed some issues where a lot of space was wasted on smaller screens. I'm not sure when this was introduced but it's fixed now. I've had to move the sidebar styles further down on the page. This should be fine as we do need to override some of the responsive styles, but I notice that the extra styles were above the responsive styles, and are now above both those and the navbar styles, making them a little more difficult to override. I feel like they should be moved below both of them, but I'm not sure what the original intention was.

goetzb commented:
@knyghty Sounds good, thank you for the fixes. The scrolling and `<br>` was a bit hacky but kind of did the job, good you got it working without the line breaks! :)

I wanted to look at the tests as well, so we could e.g. ensure the active class is applied correctly through some test.

knyghty commented:
I've removed the active class from models, all the browsers we support can use the bracketed syntax, so that and the CSS is a little cleaner now. I've also added a test that it's getting set in the correct place.

@carltongibson I think I'm happy with this now, if you could give another review.

elky commented:
Hi everyone,

I just tested current implementation and here's my thoughts:

- `float: left` thing is weak - it breaks layout in `contrib.admindocs` ([screenshot](https://user-images.githubusercontent.com/209663/80214316-4b270280-8643-11ea-8599-cc1411bf52bf.png)) and requires to hardcode css values like it's done for `messagelist`
- Static sidebar eats space so horizontal scrollbar appears for some views on small screens ([screenshot](https://user-images.githubusercontent.com/209663/80214469-904b3480-8643-11ea-8cd5-9302f357beb5.png)). It would be ok if it scrolls inside container but common scroll (that scrolls whole document) looks buggy
- Please check internationalization issues ([screenshot](https://user-images.githubusercontent.com/209663/80215518-43685d80-8645-11ea-935e-cd0eb1515696.png))
- I think app list html should be the same for index page and for sidebar. To avoid code duplication  
- When I log out I still see sidebar which is now useless
- I'm not sure it's ok to have sidebar on index page in terms of content duplication 
- Log in form is broken ([screenshot](https://user-images.githubusercontent.com/209663/80216743-351b4100-8647-11ea-8c30-e041f0c63cf4.png)) 
- In terms of UX I really want to be able to hide sidebar when it interferes me

<hr>

I tried to play with my own solution and here's demo:
http://demo.elky.me/ (login / password: demo / demo)

It's based on flex container so that all dynamic content always placed on the right. 
```
-------------
|   |       |
|   |       |
|   |       |
-------------
```
It also has a toggler which 'remembers' sidebar state with localStorage.
The issue I still can't resolve is horizontal scroll on small screens (it will probably require another container)

@knyghty @goetzb feel free to borrow code from my [solution](https://github.com/elky/django-admin-demo/pull/2/files) to simplify current implementation

elky commented on django/contrib/admin/sites.py:
```diff
@@ -302,14 +304,18 @@ def each_context(self, request):
+            'nav_sidebar_enabled': self.enable_nav_sidebar,
         }
+        if self.enable_nav_sidebar is True:
+            context['nav_app_list'] = self.get_app_list(request)
```
There is variable called `available_apps` that keeps app list and available across whole admin ;)

knyghty commented:
@elky, thanks for that. I'll have a look over the weekend but for now, your demo has no navbar for me, can you check?

elky commented:
@knyghty it's available for all pages except login, logout and index. And by default it's collapsed. E.g. http://demo.elky.me/authors/author/

knyghty commented:
Oh, I see it now, the button is just not super visible to me, thanks.

knyghty commented:
Thanks @elky, this looks great, I've adopted your changes wholesale with a few changes:

- Since the sidebar toggle button wasn't easy for me to see even when I was looking for it, I've defaulted the sidebar to open, but it will still preserve the state between reloads.
- Since there's already an admin component called sidebar, I've renamed things to nav-sidebar or similar.
- I've removed the media query. When the sidebar is closed, it looks good on all screen sizes. When it's open, it doesn't look great on mobile, but this seems preferable to the old UX where you'd have to go back to the index to go to another page, and when you're on mobile and possibly a bad connection, this seems like a win to me. There are some sizes with horizontal scroll bars, but this is the case when there is a breakpoint, too.
- I've added some of @goetzb's current page changes. I moved the aria attribute to the link itself. The spec doesn't specifically say where it should go, but all the examples I found have it on an anchor tag, so I've put it there and used classes for the styling.

I haven't tested every possible combination of page, screen size, language, etc. so more eyes are always welcome.

jdufresne commented on django/contrib/admin/templates/admin/app_list.html:
```diff
@@ -0,0 +1,36 @@
+            {% if model.add_url %}
+                <td><a href="{{ model.add_url }}" class="addlink">{% trans 'Add' %}</a></td>
+            {% else %}
+                <td>&nbsp;</td>
```
I removed these throughout the existing admin code in #12795.

elky commented on django/contrib/admin/static/admin/css/nav_sidebar.css:
```diff
@@ -0,0 +1,106 @@
+html, body {
+    height: 100%;
+}
```
> I would expect this file to only contain the CSS necessary for the sidebar itself

These changes is required to make sidebar work properly, so this is why they placed in nav_sidebar.css which includes to template if sidebar is enabled. It looks OK to me that some component adds/overrides own/global CSS rules with its own CSS file.

jdufresne replied:
I understand that, but why not apply them to the existing CSS? I think we should so that the CSS between the sidebar and non-sidebar admin version remain closer in rendering logic such that it is less likely that tweaking one version will be a detriment to the other.

knyghty replied:
The original idea here was to make sure if you disabled the navbar, everything is backwards compatible as it was before. However, I think you're right in that it'd be better to keep them as closely aligned as possible. I'll see what I can move over without breaking things for those with the sidebar disabled.

Related to this, since there's no longer any media queries, I've also moved the `<link>` tag further up the page, above the responsive CSS and above the extra styles block, so it will be easier for users to override these styles.

knyghty replied:
This is done. Everything seems ok with and without, but there are a lot of combinations.

knyghty commented on django/contrib/admin/static/admin/js/nav_sidebar.js:
```diff
@@ -0,0 +1,22 @@
+let main = document.getElementById('main');
```
This might just be personal taste but I find IIFEs to be quite ugly and a little hacky. If we don't want to pollute the global scope, and now that we have better language features available, I think a better solution would be to use `let` and put it in a block scope.

However, our ESLint setup doesn't seem to want us using ES6, although since https://github.com/django/django/pull/12502 we should be able to use it. I don't see the eslint config in the repo (but I didn't search everywhere). I assume it's in Jenkins? Could this be changed now? For such a small change I will probably just relent and use `var` and the IIFE though.

knyghty replied:
Nevermind, the config was in an extremely obvious place and somehow I still missed it.

I've changed the eslint config to allow ES6 and have gone with my solution here. However, if you think it's better as an IIFE I will change it.

jdufresne replied:
Using the ES6 blocks is good idea. You may be interested in PR #12767 which aims to do this throughout the existing code.

knyghty replied:
It's probably a good idea to merge #12767 first, as then I'll need to make a small change to the JS here.

carltongibson replied:
Grrr. Better look at #12767 then...  😀

carltongibson replied:
#12767 is in, so you can update here.

knyghty replied:
Thanks Carlton, I've sorted this out now.

knyghty commented:
Sorry about the commit spam, my editor was using the wrong autoformatting settings and kept overwriting the changes.

The main issues should be fixed now, but I'm going to have a look later at what CSS I can move into the base CSS without it breaking anything.

knyghty commented on django/contrib/admin/templates/admin/base.html:
```diff
@@ -86,8 +98,13 @@
 <!-- END Container -->
 
+{% if not is_popup and is_nav_sidebar_enabled %}
+    <script src="{% static 'admin/js/nav_sidebar.js' %}"></script>
```
Yes, seems to be good, thanks.

knyghty commented on django/contrib/admin/static/admin/js/nav_sidebar.js:
```diff
@@ -0,0 +1,22 @@
+    main.classList.add('shifted');
+} else {
+    main.classList.remove('shifted');
+}
```
Done, thanks

knyghty commented on django/contrib/admin/templates/admin/app_list.html:
```diff
@@ -0,0 +1,36 @@
+                <td><a href="{{ model.admin_url }}" class="changelink">{% trans 'Change' %}</a></td>
+            {% else %}
+                <td>&nbsp;</td>
+            {% endif %}
```
Yeah, this seems better than sending useless data, have fixed this.

knyghty replied:
There is a small issue with this approach, as the empty tds take up a little space in the sidebar, I'll take another quick look at it.

knyghty replied:
Fixed now.

goetzb commented:
@knyghty Thank you for integrating all the different changes. I like to have to have the new admin sidebar open by default.

While I'm not the biggest fan of the more subtle CSS active markup, it still does the job.

When I was quickly testing the current code from the pull request and your branch, I've noticed a JavaScript error on pages where no sidebar is present (e.g. admin homepage, app list page). To me, **it looks like the problem is that `let toggleNavSidebar = document.getElementById('toggle-nav-sidebar');` sets `toggleNavSidebar` to `null`** on such pages, so subsequent operations with `toggleNavSidebar` cause an **exception**:

> Uncaught TypeError: Cannot read property 'addEventListener' of null at nav_sidebar.js:11

Maybe it's a problem of my local test project, but I can see this behaviour for example on page `/admin/auth/`.  **It appears that the template `admin/index.html` causes this issue by removing the code for the sidebar with `{% block nav-sidebar %}{% endblock %}`**, then of course the DOM element with the ID `toggle-nav-sidebar` can't be found.

This code overwrites the `nav-sidebar` block in `admin/base.html`:
```django
          {% block nav-sidebar %}
              {% include "admin/nav_sidebar.html" %}
          {% endblock %}
```

**Adding `if (toggleNavSidebar !== null) { ... }` around `toggleNavSidebar.addEventListener('click', function() { ... });` fixes this for me.**

I've committed the change to my own branch on https://github.com/goetzb/django/commit/a95aa9269fdf1135f2059e5d6f68663dcce62d3e *(Not sure doing it this way is helpful, but I'm hesitating adding directly to the pull request or making a pull request to your branch, as I'm at this point not sure whether my local environment is messed up somehow or whether it's a general issue.)*

knyghty commented:
@goetzb good spot, didn't catch that, it does happen everywhere. I've put the if around the whole of the code since we don't need to run any of it on non-sidebar pages.

goetzb commented:
@knyghty Yes, sounds sensible. Another option would be to only load the JS and CSS on the relevant pages, but then we would need to add two more blocks in the templates - not sure that's worth it, as the CSS and JS files are pretty small (and would mostly probably be cached as well).

knyghty commented:
RTL seemed to work okay with that change, but it was still on the left. I went with a more extensive solution to get it on the right, which I'm assuming is better since that's what the rest of the admin does.

carltongibson commented:
Hi all. I'm taking another look at this again now. I just want to say a quick _Thank You_ to you all for the great collaboration on this ticket. 🥇

knyghty commented:
I think I'm happy with this state of this now, although I've said that before, so maybe there's still some more feedback 😄 

Should probably add co-author tags for this when it's squashed.

carltongibson commented:
Hi @knyghty. I'm just working through @jdufresne's clean up PRs, then this.

jdufresne commented on django/contrib/admin/templates/admin/app_list.html:
```diff
@@ -0,0 +1,36 @@
+        </div>
+    {% endfor %}
+{% else %}
+    <p>{% trans "You don't have permission to edit anything." %}</p>
```
```suggestion
    <p>{% translate "You don’t have permission to edit anything." %}</p>
```

We should keep the Unicode "Right Single Quotation Mark". It was deliberately added in 42b9a23267f14be39b9b00958e18d5746783208e.

Throughout this file, (multiple lines), the `{% trans %}` block should change to `{% translate %}` to keep commit 35d36d946272bed06a3d7c7cd4e5b71b613e7a4f.

knyghty replied:
Thanks, there were also some other changes missing from this so I've taken the fresh code from master and made my changes over the top of that.

carltongibson commented:
@knyghty I squashed and rebased. I didn't add `Co-authored-by`s yet.

carltongibson commented on docs/releases/3.1.txt:
````diff
@@ -77,6 +77,10 @@ Minor features
 * Filters in the right sidebar of the admin changelist view now contains a link
   to clear all filters.
 
+* The admin now has a sidebar for easier navigation.
+  It is enabled by default but can be disabled by using a custom AdminSite and
+  setting ``enable_nav_sidebar`` to ``False```.
+
````
```suggestion
* The admin now has a sidebar for easier navigation. It is enabled by default
  but can be disabled by using a custom AdminSite and setting
  :ref:`~django.contrib.admin.AdminSite.enable_nav_sidebar` to ``False``.
  
```

carltongibson commented on django/contrib/admin/static/admin/js/nav_sidebar.js:
```diff
@@ -0,0 +1,22 @@
+'use strict';
+{
+    const toggleNavSidebar = document.getElementById('toggle-nav-sidebar');
+    if (toggleNavSidebar !== null) {
+        const main = document.getElementById('main');
+        let navSidebarIsOpen = localStorage.getItem('django.admin.navSidebarIsOpen');
+        if (navSidebarIsOpen === null) {
+            navSidebarIsOpen = 'true';
+        }
+        main.classList.toggle('shifted', navSidebarIsOpen === 'true');
+
+        toggleNavSidebar.addEventListener('click', function() {
+            if (navSidebarIsOpen == 'true') {
+                navSidebarIsOpen = 'false';
+            } else {
+                navSidebarIsOpen = 'true';
+            }
+            localStorage.setItem('django.admin.navSidebarIsOpen', navSidebarIsOpen);
+            main.classList.toggle('shifted');
+        });
+    }
+}
```
I think a selenium test is needed for this.

knyghty replied:
I tried to add some selenium tests, but ran into a problem. It seems like even if the navbar is closed and not visible to the user, `is_displayed()` still returns `True`. It seems to be this issue: https://github.com/SeleniumHQ/selenium/issues/4689

I think we might be able to get around this by adding `display: none` in the CSS, it doesn't seem to break anything, but I'm wary of changing the CSS at this point to get a test to work.

carltongibson replied:
Hmmm. 🤔

How about a QUnit test in `js_tests` then. We can assume local storage works. 
Just testing that the nav bar toggles, and sets to and reads from local storage correctly.

(Obviously it does, but at least we have a regression test then.) 

?

knyghty replied:
I had a look at this, but I'm struggling with QUnit. Maybe I'm missing something. I added a fixture into tests.html and the relevant scripts:

```html
<script type="text/html" id="sidebar">
  <div class="sticky toggle-nav-sidebar" id="toggle-nav-sidebar"></div>
  <nav class="sticky" id="nav-sidebar"></nav>
  <div id="main"></div>
</script>

<script src="../django/contrib/admin/static/admin/js/nav_sidebar.js" data-cover></script>
<script src="./admin/NavSidebar.test.js"></script>
```

And a simple test:
```javascript
/* global QUnit, NavSidebar */
'use strict';

QUnit.module('admin.NavSidebar', {
    beforeEach: function() {
        const $ = django.jQuery;
        $('#qunit-fixture').append($('#sidebar').text());
    }
});

QUnit.test('sets local storage', function(assert) {
    const $ = django.jQuery;
    assert.equal(localStorage.getItem('django.admin.navSidebarIsOpen'), 'true');
    $('#toggle-nav-sidebar').click();
    assert.equal(localStorage.getItem('django.admin.navSidebarIsOpen'), 'false');
});
```

And I get this:
```
admin.NavSidebar - sets local storage...ERROR
>> Message: null
>> Actual: null
>> Expected: "true"
>> at Object.<anonymous> (file:///.../django/js_tests/admin/NavSidebar.test.js:13:12)

>> Message: null
>> Actual: null
>> Expected: "false"
>> at Object.<anonymous> (file:///.../django/js_tests/admin/NavSidebar.test.js:15:12)
```

I'm not exactly sure why. I get the feeling the script is running before the fixture is in, so the code doesn't run because it checks first if the element is on the page, but not having any idea how QUnit works, I don't know.

carltongibson replied:
Grrr. OK, the issue with the QUnit test is (I think) that the event listeners and such are not attached to the the `#sidebar` HTML when the page is loaded (when the `nav_sidebar.js` is loaded) because they're not part of the DOM at that point. 

When the test case is run, they are inserted into the DOM but we don't have an `init()` type function that we can call to (re)attach the even handlers. (cp the Select2 tests, which call Select2.init() with the added HTML)

So this...

```javascript

'use strict';

QUnit.module('admin.NavSidebar', {
    beforeEach: function() {
        const $ = django.jQuery;
        $('#qunit-fixture').append($('#sidebar').text());
        this.toggle = $("#toggle-nav-sidebar");
        this.sidebar = $("#nav-sidebar");
        this.main = $('#main');

    }
});

QUnit.test('sets local storage', function(assert) {
    const $ = django.jQuery;
    assert.ok(!this.main.hasClass('shifted'));
    $('#toggle-nav-sidebar').click();
    assert.ok(this.main.hasClass('shifted'));
});
```

Fails at the last assert, because the event handler is missing. 

I think the best bet is a selenium test that just checks the correct classes are toggled, and uses (I guess) `execute_script()` to check local storage is set. 

I don't think we need more that that. (That the CSS worked to adjust what's visible isn't something I want to use selenium to assert here... 😳)

Do you think you can add that @knyghty?

knyghty replied:
Done, seems to be working okay.

knyghty commented:
@carltongibson thanks for the review. I've fixed the issues with the change form and release notes.

As mentioned, though, I couldn't get a useful selenium test working 😦

elky commented:
Hey @knyghty! Thanks for you hard work on this ticket. I'm doing final checks now to make sure we didn't miss something

elky commented:
The only issue I see is small screens support. Especially mobiles:

<img src="https://user-images.githubusercontent.com/209663/81195429-7a842a80-8fc6-11ea-9855-2820210edb8a.gif" width="240">

I tried to play with it to improve this behaviour but failed. The problem is that `position: sticky` + `overflow: auto` has weak [support](https://github.com/w3c/csswg-drafts/issues/865).

My idea is to set horizontal scroll for `.main` container so that everything wider than it scrolls inside rather than outside. Here is [demo](http://jsfiddle.net/elky/4v8ez39p/35/embedded/html,css,result/dark/) of what I mean. But in this case sticky sidebar stops working. Feel free to play with it too [here](https://jsfiddle.net/elky/4v8ez39p/37/), maybe someone will find solution.

Another option to fix it is just cut everything wider by adding `overflow: hidden` to `.main.shifted`. I think it's ok for mobile but it's still an issue for small screens like laptops.

Third option is to disable sidebar for mobiles at all, or at least hide it by default.

knyghty commented:
There is quite a bit of jank on mobile when the sidebar is open, but I anticipate that people would rather deal with this jank and have a sidebar than have a jank-free existense and no sidebar, especially when you can just easily close the sidebar when you're not using it, which I imagine is how people would use it on mobile.

Just my opinion, though.

carltongibson commented:
So... mobile... 

1. The switch down the left takes too much real-estate I think. It make the change form noticeably more constrained, and, at that point, I'm the breadcrumbs and thinking I'd rather have the space and use those. 
2. The scrolling on mobile isn't great. The interaction I want is for the sidebar to come _on top_ modally and then disappear if I click outside. 
3. Finally, if I jump to a new view, when I select an item, I want the new page to be loaded without the sidebar visible, i.e. a change to the nav_sidebar.js logic in this case.

Given the closeness of the feature freeze, I don't think we can address all these issues, and safely review changes in time. I think we should probably drop showing the nav bar at all on the smallest devices, `@media (max-width: 767px)`. 

I know it's a nice to have but the original impetus here was on the big screen, and that's working nicely. I don't want to block that for 3.0 just for mobile, where it's not so convincing. 

Doing this would give time to work on the mobile case more completely. 

Is that a plan we can agree on?

knyghty commented:
@carltongibson Yeah, agreed. I've hidden it on smaller screens for now. I'll work on optmising it for mobile after this is merged 👍

carltongibson commented:
Super, thanks @knyghty! Great stuff. 

I just left one comment on the test (above)... (Then I think we're there. 🤞)

carltongibson commented on django/contrib/admin/static/admin/css/nav_sidebar.css:
```diff
@@ -88,3 +88,7 @@
 #nav-sidebar .current-model {
     background: #ffc;
 }
+
+#nav-sidebar + .content {
+    max-width: calc(100% - 24px);
+}
```
This can be reverted after 071efb13a2116a2bef638151f293f48642e9549c

knyghty replied:
Removed.

knyghty commented:
@carltongibson That's all done, thanks.

elky commented:
@knyghty one minor thing - in RTL mode there is no left border for both toggle and sidebar. Please check it ([screenshot1](https://user-images.githubusercontent.com/209663/81288749-ecfc1580-906d-11ea-9077-a292c8334690.png), [screenshot2](https://user-images.githubusercontent.com/209663/81288752-eff70600-906d-11ea-83f4-6a0be5a994b4.png))

knyghty commented:
@elky I've pushed a fix for that, thanks.

carltongibson commented:
buildbot, test on selenium.

felixxm commented on django/contrib/admin/templates/admin/base.html:
```diff
@@ -64,6 +67,15 @@
+          {% endblock %}
+      {% endif %}
+
+    <div class="content">
```
Can we fix indentation? and chop empty lines
```html
  <div class="main shifted" id="main">
    {% if not is_popup and is_nav_sidebar_enabled %}
      {% block nav-sidebar %}
        {% include "admin/nav_sidebar.html" %}
      {% endblock %}
    {% endif %}
    <div class="content">
	  ...
    </div>
  </div> 

```

knyghty replied:
I do prefer this indentation, but it wouldn't be consistent with the rest of the file. Admittedly, the rest of the file isn't very internally consistent already.

elky replied:
> 2 spaces in HTML files

The problem is that almost all current HTML files have 4 spaces. I'm suggesting to move this to separate task.

felixxm replied:
`Don’t waste time doing unrelated refactoring of existing code to conform to this style.`..., but we should follow this style in new lines.

knyghty replied:
Ok, I had to change a fair bit as there's a new element that stuff is inside, but I've tried not to change anything unnecessary. I've also fixed the indentation in `app_list.html` since that's a new file.

felixxm commented on docs/releases/3.1.txt:
```diff
@@ -77,6 +77,10 @@ Minor features
   to clear all filters.
 
+* The admin now has a sidebar on larger screens for easier navigation.
+  It is enabled by default but can be disabled by using a custom AdminSite and
```
`AdminSite` -> ``` ``AdminSite`` ```

felixxm commented on tests/admin_views/test_nav_sidebar.py:
```diff
@@ -0,0 +1,97 @@
+
+
+site_with_sidebar = AdminSiteWithSidebar(name="test_with_sidebar")
+site_without_sidebar = AdminSiteWithoutSidebar(name="test_without_sidebar")
```
Single quotes.

elky commented:
@knyghty I found a bug in Safari and fixed it here https://github.com/elky/django/commit/1ab40eda4b60c065073b182b4c1b2e063cd6bb9a. Please add it to the PR.

knyghty commented:
Added that, thanks @elky

felixxm commented:
Thanks y'all :confetti_ball: 

I pushed small edits and moved cleanups to the separate commit.

felixxm commented:
I attached a screenshoot from `admin_inlines.tests.SeleniumTests.test_inline_formset_error`:
![Bez nazwy2](https://user-images.githubusercontent.com/2865885/81345053-193c8400-90b8-11ea-8d41-82532f19984d.png)

knyghty commented:
It's strange, I can only replicate this locally when running selenium headless, it doesn't seem to happen when I ran without `--headless` so it's a little hard to debug. I'll try to investigate tomorrow.

knyghty commented:
Blargh! It looks like when selenium is running headless it's running in the 768 - 1024px range. I can fix this locally by using the 1024px media query (there are still some issues at this size, though fixable by closing the navbar, so this seems like a good change anyway). However, this breaks the new selenium tests for the sidebar, as the sidebar toggle is hidden at this size by the media query.

@carltongibson, @felixxm, any suggestions?

knyghty commented:
Ah, sorry for the spam, I haven't used selenium much before, didn't realise I could set the viewport size per test, this should be fixed now 😌

felixxm commented:
buildbot, test on selenium.

felixxm commented:
Locally it still doesn't work for me :confused:
 - Firefox, `{'width': 1229, 'height': 1891}` :red_circle:
 - Firefox, headless, `{'width': 1366, 'height': 768}` :green_circle:
 - Chrome, `{'width': 1050, 'height': 691}` :red_circle:
 - Chrome, headless, `{'width': 800, 'height': 600}` :green_circle:

4ff675b32f9951bc4b24b7dcef9836ab199e523b fixed only headless Chrome for me.

felixxm commented:
As you can see on a screenshot this is a real issue with tabular inlines:

![obraz](https://user-images.githubusercontent.com/2865885/81540988-c0e4db00-9372-11ea-88bf-b63fffa55d7d.png)

felixxm commented:
@knyghty I reverted your fix and added two changes:

- 1st, [scroll to the left side in `test_inline_formset_error`](https://github.com/django/django/pull/12159/commits/f9fb7e1a4cbe980613b69bda7e4286cda1549ad4#diff-493bd093cf2d3c53b8f1cde87572162aR1334-R1336) because a tabular inline contains a vertical scroll. This change fixes inlines tests in all combinations (Firefox / Chrome / headless).
- 2nd, independent fix for `TextField` CSS that was broken with a large `48em` width. I changed it to match `TextField` width for `.colMS .aligned .vLargeTextField`. Before:
![obraz](https://user-images.githubusercontent.com/2865885/81547388-1376c500-937c-11ea-897b-6f35277671e0.png)
After:
![obraz](https://user-images.githubusercontent.com/2865885/81547272-ed512500-937b-11ea-8551-0f96d88224b9.png)

felixxm commented:
buildbot, test on selenium.

knyghty commented:
Thanks for that, I was looking at it a bit earlier. It seems like it'll still be an issue if your inline has even more fields, but this is an issue without the sidebar. The tabular inlines can get very large and cause horizontal scrolling, and in this case the sidebar ends up covering it as you scroll, but it can be mitigated by just closing it, so I think it's alright from the user's perspective.

felixxm commented:
> Thanks for that, I was looking at it a bit earlier. It seems like it'll still be an issue if your inline has even more fields, but this is an issue without the sidebar. The tabular inlines can get very large and cause horizontal scrolling, and in this case the sidebar ends up covering it as you scroll, but it can be mitigated by just closing it, so I think it's alright from the user's perspective.

Agreed, that's why I added scrolling to the left side.

felixxm commented:
Thanks y'all :rocket: 

I reverted my last cleanup.

felixxm commented:
buildbot, test on selenium.

jdufresne commented on django/contrib/admin/templates/admin/app_list.html:
```diff
@@ -0,0 +1,40 @@
+
+{% if app_list %}
+  {% for app in app_list %}
+    <div class="app-{{ app.app_label }} module{% if app.app_url in request.path %} current-app{% endif %}">
```
The use of `request` in the template means the admin now has a dependency on `django.template.context_processors.request`. Technically, this is backwards incompatible. That doesn't particularly bother me, but perhaps should be noted in the release notes. This typically exists from the default settings, so perhaps not a problem in practice.

This should be updated in in the admin docs: https://docs.djangoproject.com/en/dev/ref/contrib/admin/#overview

As well as to the admin checks: https://github.com/django/django/blob/d51e090db2110f016dbca1d794c0d379b3df551b/django/contrib/admin/checks.py#L91-L107

(I discovered this rebasing PR #12752.)

jdufresne replied:
Addressed in PR #12906

endersonmenezes commented:
Has anyone seen this error? Was this repeated with someone during the tests?

[Stackoverflow Questio](https://stackoverflow.com/questions/63286591/django-3-1-admin-page-appearance-issue)

knyghty commented:
@endersonmenezes I believe it's just a browser cache issue.

endersonmenezes commented:
@knyghty This bug happened to me and I am researching it, and the documentation says that the new requirement is to add "django.contrib.messages.context_processors.request" to "context_processors", but the one inside "messages.conext_processors" does not exist "request".

knyghty commented:
@endersonmenezes can you point to where this is in the docs?

I believe the string should be `django.template.context_processors.request`, not `django.contrib.messages.context_processors.request`

endersonmenezes commented:
That's right, sorry I'm a beginner developer and I couldn't express myself well.

knyghty commented:
@endersonmenezes No problem. By the way, a closed PR isn't the best place to get support, you might have better luck on IRC or the mailing list: https://www.djangoproject.com/community/
